GH-Actions: Add missing labels

macho-catt opened the pull request:
Fixes #2013

### What changes did you make and why did you make them ?

  - Added gh-action to add the required missing labels on an issue if a user forgets them.
  -  The gh-action also removes a "missing label" if the required label is actually filled out. (i.e. I accidentally added the "role missing" label even though I added the "role: front end" properly. The gh-action will remove the "role missing" label and keep the proper label)
  - The gh-action also posts a comment depending on if the user added the required labels or not

### Screenshots of example issues on my fork

<details>
<summary>Issue with no labels</summary>

![Screenshot from 2021-08-03 19-03-39](https://user-images.githubusercontent.com/48004150/128110089-1bd9da05-213f-4813-8cdc-65f3bfa6614c.png)


</details>

<details>
<summary>Issue with required labels</summary>
  
![Screenshot from 2021-08-03 19-03-59](https://user-images.githubusercontent.com/48004150/128110119-cccc3ac8-0fbb-4843-b072-a3c00b57f3bc.png)


</details>

<details>
<summary>Issue with extraneous labels</summary>
  
![Screenshot from 2021-08-03 19-04-15](https://user-images.githubusercontent.com/48004150/128110131-c4849dee-87d2-414e-89e3-8a02cf01b780.png)


</details>

macho-catt committed:
Create new yml file for gh action

macho-catt committed:
Add js script to check labels

macho-catt committed:
Create JS file to check labels

macho-catt committed:
Export main function in js

macho-catt committed:
Fix incorrect path

macho-catt committed:
Remove return param

macho-catt committed:
Test console log output

macho-catt committed:
Remove return

macho-catt committed:
Remove imports

macho-catt committed:
Add checkout action

macho-catt committed:
Fix indent

macho-catt committed:
Uncomment commented code

macho-catt committed:
Test

macho-catt committed:
Add labels

macho-catt committed:
Test context object

macho-catt committed:
Add payload

macho-catt committed:
Obtain labels from object

macho-catt committed:
Check labels on issue

macho-catt committed:
Add more attributes

macho-catt committed:
Remove unwanted labels

macho-catt committed:
Add async

macho-catt committed:
Added console.log

macho-catt committed:
Fix logic

macho-catt committed:
Change to filter

macho-catt committed:
Filter

macho-catt committed:
Fix typo

macho-catt committed:
Clean function

macho-catt committed:
Add promise

macho-catt committed:
Check labels

macho-catt committed:
Add debug

macho-catt committed:
FIxed regex

macho-catt committed:
Fix regex

macho-catt committed:
Fix labels to add

macho-catt committed:
Add labels

macho-catt committed:
Fix variables

macho-catt committed:
Fix labels to add

macho-catt committed:
Fix typo

macho-catt committed:
Update try catch

macho-catt committed:
Refactored code

macho-catt committed:
Flatten set

macho-catt committed:
Clean up code

macho-catt committed:
New action

macho-catt committed:
Fix syntax

macho-catt committed:
Fix export

macho-catt committed:
Check issue creator

macho-catt committed:
Fixed syntax

macho-catt committed:
Add comment

macho-catt committed:
Add async

macho-catt committed:
Fix comments

macho-catt committed:
Fixed object

macho-catt committed:
Update template

macho-catt committed:
More template updates

alexeysergeev-cm commented on github-actions/issue-labels/check-labels.js:
```diff
@@ -0,0 +1,91 @@
+  const issueNum = context.payload.issue.number
+
+  const labels = obtainLabels()
+  const filteredLabels = await filterLabels(labels)
```
Is this function `async` at all? It seems it does not need `await` or am I missing something?

macho-catt replied:
I can remove the async. Previously it was an async function that I refactored, but I missed removing this part. Good catch.

alexeysergeev-cm commented on github-actions/issue-labels/post-labels-comment.js:
```diff
@@ -0,0 +1,77 @@
+function makeComment(labels) {
+  const issueCreator = context.payload.issue.user.login
+
+  if (labels.length === 0) {
+    return `Good job @${issueCreator} for adding the required labels to this issue.`
+  }
+
+  return `Hi @${issueCreator}. Please don't forget to add the proper labels to this issue.
+  Currently, the labels for the following are missing:
+  ${labels.map(label => ` ${LABELS_OBJ[label]} `)}
+  To add a label, take a look at Github's documentation [here](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#applying-a-label).
+  Also, don't forget to remove the "missing labels" after you add the proper labels.
+  To remove a label, the process is similar to adding a label, but you select a currently added label to remove it.`
+}
```
Is there an option we can put this in a separate `.MD` file and just pass the username to avoid touching code if the change in the message is needed? 

Can we also remove spaces in the ` ' ${LABELS_OBJ[label]}  '` on line 49 to avoid the spacing in the comment, or is that necessary?

macho-catt replied:
I can do that. It might actually be better that way. Thanks for the suggestion!

alexeysergeev-cm reviewed (commented):
Hi @macho-catt, amazing work! I like how the code is easy to read!

macho-catt committed:
Merge branch 'gh-pages' into add-missing-labels-2013

macho-catt committed:
Change async function to sync

macho-catt committed:
Add new md files for comments

macho-catt committed:
Add issue creator to comment

macho-catt committed:
Fix style

macho-catt committed:
Add description to functions

macho-catt committed:
Extract functions to a utility folder

macho-catt committed:
Add github and context object params

macho-catt committed:
Add descriptive comments

macho-catt committed:
Fix templates

macho-catt committed:
Fix template comments

macho-catt commented:
Updates based on @alexeysergeev-cm suggestions

### Additional changes

  - Changed the filterLabels function under check-labels.js to be synchronous
  - Added new MD files for comment templates, instead of only having one:
    - labels-instructions-template.md - overall comment to be posted
    - add-labels-template.md - instructions template for labels that need to be added
    - no-labels-template.md - comment if required labels are added
    - fixed spacing for the labels in the posted comment
  - Updated logic for creating a comment
  - Updated formatComment function to be more usable outside of this issue
  - Updated postComment function to be more usable outside of this issue
  - Separated the two functions above into the utils folder so other files can use them

### Some comments
Since the format comment and post comment are used in other gh-action js files, (the overall function and result are the same), I extracted them into their own files. I noticed that other js files (such as post-comment.js under pr-instructions and add-label.js under add-update-label-weekly) use them as well. They might need to be re-edited a bit, but I think it'll be good moving forward to use this. What do you think @Aveline-art ? Also, I saw two labels that don't follow the format of other labels in the same group: "Role: Administrative" and "Role: Product". Can these be changed to follow the format "role: xxxx" since that's how the other roles are formatted?

It should also be noted that if any of the size, roles, or features labels are edited / updated in the feature, the new code stemming from this PR should be noted since the code uses the pattern of how the current labels are.

### Screenshots of example issues on my fork

<details>
<summary>Updated issue with no labels</summary>

![no labels](https://user-images.githubusercontent.com/48004150/128620830-0a9dae71-f06f-46c7-a7de-85c44b06f2f1.png)


</details>

<details>
<summary>Updated issue with required labels</summary>
  
![with-labels](https://user-images.githubusercontent.com/48004150/128620838-e760d59b-f0f0-4caf-a594-b2a1a5a37d69.png)


</details>

Aveline-art commented on github-actions/utils/comments.js:
```diff
@@ -0,0 +1,41 @@
+ * @param {Number} issueNum - the issue number where the comment should be posted
+ * @param {String} comment - the comment to be posted
+ */
+ async function postComment(github, context, issueNum, comment) {
+  try {
+    await github.issues.createComment({
+      owner: context.repo.owner,
+      repo: context.repo.repo,
+      issue_number: issueNum,
+      body: comment,
+    })
+  } catch (err) {
+    throw new Error(err);
+  }
+}
+
```
I am going to veto this, since it also requires a lot of parameters, and the catch function should be customizable according to the preferred performance of the function. API calls by its nature already reduce the needed parameters as much as possible, so I think trying to reduce it even more takes away some of the flexibility of it.

macho-catt replied:
Okay, I can move it back to the original folder

Aveline-art commented on github-actions/utils/comments.js:
```diff
@@ -0,0 +1,41 @@
+ * @param {String} textToFormat - the text to be formatted. If null, use the md file provided in the path. If provided, format that text
+ * @returns {String} - returns a formatted comment to be posted on github
+ */
+ function formatComment(replacementString, placeholderString, path, textToFormat) {
+  const text = textToFormat === null ? fs.readFileSync(path).toString('utf-8') : textToFormat
+  const commentToPost = text.replace(placeholderString, replacementString)
+  return commentToPost
+}
```
I can see why this should be here, but I also hesitate since 4 required parameters seem to defeat the purpose of making an outside function. I think a utility function should be easy to understand, and does one task very well. I can also envision the need for multiple placeholders, which this does not account for. Not sure if either of those concerns can be addressed, but if not, just put this back into the main folder..

macho-catt replied:
I'll also move this back as well.

Aveline-art commented on github-actions/issue-labels/labels-instructions-template.md:
```diff
@@ -0,0 +1,13 @@
+<!-- Template for a comment to be posted after an issue is created. --->
+<!-- The issue creator's handle replaces the first placeholder and the instructions replace the second placeholder. --->
+
+Hi @${issueCreator}.
+
+${labelInstructions}
+
+Just a friendly reminder to also move the issue into the "New Issue Approval" column under the Project Board. For more info on that, check Github's documentation [here](https://docs.github.com/en/issues/organizing-your-work-with-project-boards/tracking-work-with-project-boards/adding-issues-and-pull-requests-to-a-project-board).
+Thanks!
+
+**Additional Resources:**
+Wiki: How to add status labels to issues (_WIP. Link will be updated when the wiki is done_)
+Wiki: How to create issues (_WIP. Link will be updated when the wiki is done_)
```
I think we would prefer to have two files, rather than a template file that can be adjusted with another file. The idea behind the outside md file is so that a user can clearly see what message the client is seeing (or in other words, a PM, with no programming experience, can adjust their message and get a clear idea of what their message is). Can this be moved into the other templates? It is okay if the files are very similar; it is much more important that we get a grasp of what message we are sending.

macho-catt replied:
I'll consolidate this and add it into the two files

Aveline-art commented on github-actions/issue-labels/check-labels.js:
```diff
@@ -0,0 +1,106 @@
+  let labelsToAdd = []
+
+  REQUIRED_LABELS.forEach((requiredLabel, i) => {
+    const regExp = new RegExp(`\\b${requiredLabel}\\b`, 'g')
+    const isLabelPresent = labels.some(label => regExp.test(label))
+
```
I've noticed an issue with the role vs Role labels. I think the [i modifier ](https://www.w3schools.com/jsref/jsref_regexp_i.asp)would help with that? I am not sure if there could be another edge case, but if so, let me know so I can make adjustments.

macho-catt replied:
I could try that, that might actually make it easier

Aveline-art commented on github-actions/issue-labels/check-labels.js:
```diff
@@ -0,0 +1,106 @@
+  const repo = context.payload.repository.name
+
+  // Use a flattened set to ensure each label is unique
+  const labels = [...new Set([
+    ...labelsToAdd,
+    ...currentLabels
+  ])]
+
```
Not sure why the set object needs to be unpacked. Does the API not take a set object?

macho-catt replied:
I only wanted to use the Set object to clear duplicates when I merge the two arrays, and initially when I wrote this and passed it on I wanted to retain it as an array so I can use any properties that come with it (for example, I can use map with an array but with a set I have to use the spread syntax first to transform it to an array). I also wanted to remove any unforeseen effects in using a Set object, so to make my life easier I made it an array.

I could change it back to just a Set if there's a need for that, but otherwise I think using it as an array will make it easier to use.

Aveline-art reviewed (changes requested):
Great work on this! Just a couple of aesthetic changes for this one. Once those are addressed, I can start testing these changes.

macho-catt committed:
Remove utils/comments file

macho-catt committed:
Revert postComment function

macho-catt committed:
Consolidate template files

macho-catt committed:
Change mult params to single obj for readability

macho-catt committed:
Update function info

macho-catt committed:
Update function logic for makeComment

macho-catt committed:
Update regex

macho-catt committed:
Remove unused md file

macho-catt commented:
### Additional changes

  - Removed the utils/comment.js file and added the two functions there back to post-labels-comment.js
  - Removed the labels-instructions-template.md file and added the body from that into the two other md files
  - Changed the input parameters for formatComment to take an object instead of four parameters.
    - Pros and Cons of using multiple parameters vs an options object: https://stackoverflow.com/questions/12826977/multiple-arguments-vs-options-object
 - Added i flag to regex to ignore character case
 - Updated logic for makeComment function to accommodate the changes for formatComment

Aveline-art commented on github-actions/issue-labels/add-labels-template.md:
```diff
@@ -0,0 +1,18 @@
+<!-- Template for a comment that requires the user to add any missing labels. --->
+
+Hi @${issueCreator}.
+
+Please don't forget to add the proper labels to this issue. Currently, the labels for the following are missing:
+${labels}
+
+To add a label, take a look at Github's documentation [here](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#applying-a-label).
+
```
I just realized that there is a use case for not using a label (a PM or designer might not know the size of a development issue, for example, and vice versa). In such a case, we would not want to encourage adding a label and hoping for the best, but rather allow someone from the team with the required experience to add the labels instead. For now, let's comment out the function that posts this template (and leave the other template as is) and make a new issue for product to brainstorm the desired behavior and message

Aveline-art replied:
@macho-catt Just comment out the code and this will be ready to merge :) Great work!

macho-catt replied:
@Aveline-art so just comment out lines 5-8 of add-labels-template.md (and accompanying code that maps the labels into this template) until we figure out what PMs want?

Aveline-art replied:
Mmm, do it from the js side, so add-labels-template is never posted in the first place. I don't think the md file needs to be commented out at all?

macho-catt replied:
Ok, so if no labels were added in the first place, don't have the bot comment anything and just add the three missing labels. Is that correct?

Aveline-art replied:
Yep. The missing labels should suffice until the message can be tailored for our optimal workflow.

macho-catt committed:
Remove the action to post a comment because of an unforeseen use case PR #2060

macho-catt commented:
### Additional changes

  - Commented out the code where the action will post when the required labels are not added. The action will end (instead of posting something)
  - Added a comment in the code that details why we did this

Aveline-art reviewed (commented):
LGTM!

Aveline-art commented:
The GHA works beautifully! Thanks for making this!
